File: hawkeye_widget/src/main/java/com/doyou/cv/widget/HalfCircleProView.java
package com.doyou.cv.widget;

import android.animation.ValueAnimator;
import android.content.Context;
import android.content.res.TypedArray;
import android.graphics.Bitmap;
import android.graphics.BitmapFactory;
import android.graphics.Canvas;
import android.graphics.Color;
import android.graphics.LinearGradient;
import android.graphics.Paint;
import android.graphics.RectF;
import android.graphics.Shader;
import android.util.AttributeSet;
import android.view.View;
import android.view.animation.LinearInterpolator;

import com.dongni.tools.DensityUtil;
import com.doyou.cv.R;
import com.doyou.cv.utils.FormatUtil;
import com.doyou.cv.utils.LogUtil;

import androidx.annotation.Nullable;

/**
 * 油表控件
 */
public class HalfCircleProView extends View {
    /**
     * 圆心x坐标
     */
    private float centerX;
    /**
     * 圆心y坐标
     */
    private float centerY;

    /**
     * 扇形所在矩形
     */
    private RectF rectF = new RectF();

    /**
     * 进度动画
     */
    private ValueAnimator progressAnimator;

    /**
     * 动画执行时间
     */
    private int duration = 500;
    /**
     * 动画延时启动时间
     */
    private int startDelay = 500;

    /**
     * 圆形的画笔
     *
     * @param context
     */
    private Paint mProgressPaint;

    /**
     * 百分比的画笔
     */
    private Paint mTxtPaint;
    /**
     * 圆形的宽度
     *
     * @param context
     */
    private int mPaintWidth;

    /**
     * @param context
     */
    private Bitmap mBitmap;

    private Context mContext;

    /**
     * 进度条的半径
     */
    private int mRadius;
    /**
     * 内环的半径
     */
    private int mInRadius;
    private int[] mGradientColors = {0xffffbf05, 0xFFFF6953};
    private float mProgress;
    private float mTempProgress;

    /**
     * 百分比文本的大小
     */
    private int mTxtSize;
    /**
     * 百分比的颜色
     */
    private int mTxtColor;

    public HalfCircleProView(Context context) {
        this(context, null, 0);
    }

    public HalfCircleProView(Context context, @Nullable AttributeSet attrs) {
        this(context, attrs, 0);
    }

    public HalfCircleProView(Context context, @Nullable AttributeSet attrs, int defStyleAttr) {
        super(context, attrs, defStyleAttr);
        mContext = context;
        init(attrs);
    }

    private void init(AttributeSet attrs) {
        getAttr(attrs);
        mProgressPaint = getPaint(mPaintWidth, Color.RED);
        initTextPaint();
        initBitmap();

    }

    private Paint getPaint(int strokeWidth, int color) {
        Paint paint = new Paint(Paint.ANTI_ALIAS_FLAG);
        paint.setStrokeWidth(strokeWidth);
        paint.setColor(color);
        paint.setAntiAlias(true);
        paint.setStrokeCap(Paint.Cap.ROUND);
        paint.setStyle(Paint.Style.STROKE);

        return paint;
    }

    private void initTextPaint() {
        mTxtPaint = new Paint(Paint.ANTI_ALIAS_FLAG);
        mTxtPaint.setTextSize(mTxtSize);
        mTxtPaint.setColor(mTxtColor);
        mTxtPaint.setTextAlign(Paint.Align.CENTER);
        mTxtPaint.setAntiAlias(true);
    }

    public void setProgress(float progress) {
        this.mProgress = progress;
        initAnimation();
    }

    public void setGradientColors(int[] colors) {
        this.mGradientColors = colors;
    }


    private void getAttr(AttributeSet attrs) {
        TypedArray typedArray = mContext.obtainStyledAttributes(attrs, R.styleable.HalfCircleProView);
        mPaintWidth = typedArray.getDimensionPixelOffset(R.styleable.HalfCircleProView_strokeWidth, DensityUtil.dp2px(2));
        mTxtColor = typedArray.getColor(R.styleable.HalfCircleProView_txtColor, getResources().getColor(R.color.txt_black));
        mTxtSize = typedArray.getDimensionPixelSize(R.styleable.HalfCircleProView_txtSize, DensityUtil.dp2px(12));
        typedArray.recycle();
    }

    private void initBitmap() {
        mBitmap = BitmapFactory.decodeResource(getResources(), R.drawable.icon_half_cicle);
        LogUtil.logD("201810301418", "initBitmap");
        mInRadius = mBitmap.getWidth() / 2;
        mRadius = mInRadius + mPaintWidth /*+ dp2px(1)*/;
    }


    @Override
    protected void onSizeChanged(int w, int h, int oldW, int oldH) {
        super.onSizeChanged(w, h, oldW, oldH);
        centerX = w / 2;
        centerY = h;
        rectF.set(centerX - mRadius, centerY - mRadius, centerX + mRadius, centerY + mRadius);
        LogUtil.logD("201810301418", "onSizeChanged");
    }

    @Override
    protected void onMeasure(int widthMeasureSpec, int heightMeasureSpec) {
        int widthSize = mRadius * 2 + mPaintWidth + getPaddingLeft() + getPaddingRight();
        int heightSize = mRadius + mPaintWidth + getPaddingTop() + getPaddingBottom();
        int width = resolveSize(widthSize, widthMeasureSpec);
        int height = resolveSize(heightSize, heightMeasureSpec);
        LogUtil.logD("onMeasure", "width = " + width + "->height = " + height);
        setMeasuredDimension(width, height);
    }

    @Override
    protected void onDraw(Canvas canvas) {
        super.onDraw(canvas);
        LinearGradient sweepGradient = new LinearGradient(centerX - mRadius, centerY, centerX + mRadius, centerY, mGradientColors, null, Shader.TileMode.MIRROR);
        mProgressPaint.setShader(sweepGradient);
        canvas.drawBitmap(mBitmap, centerX - mInRadius, centerY - mInRadius, mProgressPaint);
        canvas.drawText(FormatUtil.formatNumToPercentByTwoPoint(mProgress), centerX, centerY, mTxtPaint);
        mTempProgress = mProgress == 0 ? 0.1f : mProgress;
        float sweepAngele = mTempProgress * 180 / 100;
        canvas.drawArc(rectF, 180, sweepAngele, false, mProgressPaint);
    }

    private void initAnimation() {
        progressAnimator = ValueAnimator.ofFloat(0, mProgress);
        progressAnimator.setDuration(duration);
        progressAnimator.setStartDelay(startDelay);
        progressAnimator.setInterpolator(new LinearInterpolator());
        progressAnimator.addUpdateListener(valueAnimator -> {
            float value = (float) valueAnimator.getAnimatedValue();
            mProgress = value;
            invalidate();
        });
        progressAnimator.start();
    }

}

File: lib-widget/src/main/java/com/doyou/cv/widget/HalfCircleProView.kt
package com.doyou.cv.widget

import android.animation.ValueAnimator
import android.content.Context
import android.graphics.Bitmap
import android.graphics.BitmapFactory
import android.graphics.Canvas
import android.graphics.Color
import android.graphics.LinearGradient
import android.graphics.Paint
import android.graphics.RectF
import android.graphics.Shader
import android.util.AttributeSet
import android.view.View
import android.view.animation.LinearInterpolator
import androidx.core.content.ContextCompat
import com.doyou.cv.R
import com.doyou.cv.utils.FormatUtil.formatNumToPercentByTwoPoint
import com.doyou.cv.utils.LogUtil.logD
import com.doyou.tools.DensityUtil.dp2px

/**
 * 油表控件
 */
class HalfCircleProView @JvmOverloads constructor(
    private val mContext: Context,
    attrs: AttributeSet? = null,
    defStyleAttr: Int = 0
) : View(
    mContext, attrs, defStyleAttr
) {
    /**
     * 圆心x坐标
     */
    private var centerX = 0f

    /**
     * 圆心y坐标
     */
    private var centerY = 0f

    /**
     * 扇形所在矩形
     */
    private val rectF = RectF()

    /**
     * 进度动画
     */
    private var progressAnimator: ValueAnimator? = null

    /**
     * 动画执行时间
     */
    private val sDuration = 500

    /**
     * 动画延时启动时间
     */
    private val sStartDelay = 500

    /**
     * 圆形的画笔
     */
    private var mProgressPaint: Paint? = null

    /**
     * 百分比的画笔
     */
    private var mTxtPaint: Paint? = null

    /**
     * 圆形的宽度
     */
    private var mPaintWidth = 0

    /**
     * 位图
     */
    private var mBitmap: Bitmap? = null

    /**
     * 进度条的半径
     */
    private var mRadius = 0

    /**
     * 内环的半径
     */
    private var mInRadius = 0
    private var mGradientColors = intArrayOf(-0x40fb, -0x96ad)
    private var mProgress = 0f
    private var mTempProgress = 0f

    /**
     * 百分比文本的大小
     */
    private var mTxtSize = 0

    /**
     * 百分比的颜色
     */
    private var mTxtColor = 0

    init {
        init(attrs)
    }

    private fun init(attrs: AttributeSet?) {
        getAttr(attrs)
        mProgressPaint = getPaint(mPaintWidth)
        initTextPaint()
        initBitmap()
    }

    private fun getPaint(strokeWidth: Int): Paint {
        val paint = Paint(Paint.ANTI_ALIAS_FLAG)
        paint.strokeWidth = strokeWidth.toFloat()
        paint.color = Color.RED
        paint.isAntiAlias = true
        paint.strokeCap = Paint.Cap.ROUND
        paint.style = Paint.Style.STROKE
        return paint
    }

    private fun initTextPaint() {
        mTxtPaint = Paint(Paint.ANTI_ALIAS_FLAG)
        mTxtPaint!!.textSize = mTxtSize.toFloat()
        mTxtPaint!!.color = mTxtColor
        mTxtPaint!!.textAlign = Paint.Align.CENTER
        mTxtPaint!!.isAntiAlias = true
    }

    fun setProgress(progress: Float) {
        mProgress = progress
        initAnimation()
    }

    fun setGradientColors(colors: IntArray) {
        mGradientColors = colors
    }

    private fun getAttr(attrs: AttributeSet?) {
        val typedArray = mContext.obtainStyledAttributes(attrs, R.styleable.HalfCircleProView)
        mPaintWidth =
            typedArray.getDimensionPixelOffset(R.styleable.HalfCircleProView_strokeWidth, dp2px(2f))
        mTxtColor = typedArray.getColor(
            R.styleable.HalfCircleProView_txtColor,
            ContextCompat.getColor(context, R.color.txt_black)
        )
        mTxtSize =
            typedArray.getDimensionPixelSize(R.styleable.HalfCircleProView_txtSize, dp2px(12f))
        typedArray.recycle()
    }

    private fun initBitmap() {
        mBitmap = BitmapFactory.decodeResource(resources, R.drawable.icon_half_cicle)
        logD("201810301418", "initBitmap")
        mInRadius = mBitmap?.width!! / 2
        mRadius = mInRadius + mPaintWidth /*+ dp2px(1)*/
    }

    override fun onSizeChanged(w: Int, h: Int, oldW: Int, oldH: Int) {
        super.onSizeChanged(w, h, oldW, oldH)
        centerX = (w / 2).toFloat()
        centerY = h.toFloat()
        rectF[centerX - mRadius, centerY - mRadius, centerX + mRadius] = centerY + mRadius
        logD("201810301418", "onSizeChanged")
    }

    override fun onMeasure(widthMeasureSpec: Int, heightMeasureSpec: Int) {
        val widthSize = mRadius * 2 + mPaintWidth + paddingLeft + paddingRight
        val heightSize = mRadius + mPaintWidth + paddingTop + paddingBottom
        val width = resolveSize(widthSize, widthMeasureSpec)
        val height = resolveSize(heightSize, heightMeasureSpec)
        logD("onMeasure", "width = $width->height = $height")
        setMeasuredDimension(width, height)
    }

    override fun onDraw(canvas: Canvas) {
        super.onDraw(canvas)
        val sweepGradient = LinearGradient(
            centerX - mRadius,
            centerY,
            centerX + mRadius,
            centerY,
            mGradientColors,
            null,
            Shader.TileMode.MIRROR
        )
        mProgressPaint!!.shader = sweepGradient
        canvas.drawBitmap(mBitmap, centerX - mInRadius, centerY - mInRadius, mProgressPaint)
        canvas.drawText(formatNumToPercentByTwoPoint(mProgress), centerX, centerY, mTxtPaint)
        mTempProgress = if (mProgress == 0f) 0.1f else mProgress
        val sweepAngele = mTempProgress * 180 / 100
        canvas.drawArc(rectF, 180f, sweepAngele, false, mProgressPaint)
    }

    private fun initAnimation() {
        progressAnimator = ValueAnimator.ofFloat(0f, mProgress)
        progressAnimator?.apply {
            duration = sDuration.toLong()
            startDelay = sStartDelay.toLong()
            interpolator = LinearInterpolator()
            addUpdateListener { valueAnimator: ValueAnimator ->
                val value = valueAnimator.animatedValue as Float
                mProgress = value
                invalidate()
            }
            start()
        }
    }
}